frame/staking/src/lib.rs

```diff
@@ -3341,19 +3341,45 @@ impl<T: Config> Module<T> {
 	}
 }
 
-impl<T: Config> sp_election_providers::ElectionDataProvider<T::AccountId, T::BlockNumber>
+impl<T: Config> frame_election_provider_support::ElectionDataProvider<T::AccountId, T::BlockNumber>
 	for Module<T>
 {
-	fn desired_targets() -> u32 {
-		Self::validator_count()
+	fn desired_targets() -> data_provider::Result<(u32, Weight)> {
+		Ok((Self::validator_count(), <T as frame_system::Config>::DbWeight::get().reads(1)))
 	}
 
-	fn voters() -> Vec<(T::AccountId, VoteWeight, Vec<T::AccountId>)> {
-		Self::get_npos_voters()
+	fn voters(
+		maybe_max_len: Option<usize>,
+	) -> data_provider::Result<(Vec<(T::AccountId, VoteWeight, Vec<T::AccountId>)>, Weight)> {
+		// NOTE: reading these counts already needs to iterate a lot of storage keys, but they get
+		// cached. This is okay for the case of `Ok(_)`, but bad for `Err(_)`, as the trait does not
+		// report weight in failures.
```

Comment:
which cache ?
AFAIK we don't cache what is the next key of some key. Thus calling `sp_io::next_key(k)` 2 times will search 2 times what is the next key for `k` in the backend.

Sidenote: I opened an issue about my worry on not counting the next_key calls: https://github.com/paritytech/substrate/issues/8363

Reply:
> which cache ?

Overlay. 

And based I think the note is incorrect. We first read th count here, which does not cache the `next_key`, thus next time when we read the whole map again, everything need to be fetched again? That's bad, but sadly not much that we can do for now. Maybe we should start storing the count of validator and nominator maps in storage.

Reply:
About the implementation, given that the cost of decode is much cheaper than storage access, and the fact that all in all we expect this function to return error very rarely, it is a better option to not count the keys at all: we just iter and collect all of them, and if they are more than what we expected, we drop the values. 

But we can also re-think this later. This is a once-per-day on-Initialize event, not like we can do anything about it at this point if it takes more than the block weight 🤷

Reply:
I mixed myself, I was thinking `storage_map.iter().count()` would only call `next_keys` but indeed it doesn't it does read the value even if it makes no use of it.

ok forget what I said, indeed the value there are cached.

Reply:
ahh interesting, okay then issue is solved from my perspective. Thanks